File: simulate.py
# -*- Coding: utf-8 -*-

"""
現在botは2人用を想定した設計で制作。
そのため、2人用に特化したreversi_duelを用いると(ほんの少し)動作が早くなります。

GUIの方にreversi_duelは未対応。
"""

import reversi as reversi
import bot_methods as bot
import load

def main():
    # reversi program
    
    ## game initialize
    game = reversi.Reversi(
        load.parse_config('simulate_config.json'),
        area = 8
    )

    game.game_init()

    # game main loop
    while True:
        # print('PLAYER {} TURN'.format(game.ORDER.index(game.turn)+1))
        pos = game.turn.rogic(game)
        print('{} put position -> ({}, {})'.format(game.stone_num, *pos))

        game.put_stone(pos, game.turn)
        
        game.nextturn()

        game.putlist = game.make_putlist(game.turn)

        pass_player = game.pass_check()
        if len(pass_player) == len(game.ORDER):
            # ゲーム終了
            game.end()
            print(game.score)
            return
        else:
            for p in pass_player:
                # print('\n\nGAME INFO\nPLAYER {}, PASS\n'.format(game.ORDER.index(game.turn)+1))
                pass

    return

import time
if __name__ == '__main__':
    start = time.time()
    main()
    elapsed_time = time.time() - start
    print("elapsed time is {} [sec]".format(elapsed_time))

File: simulate.c
/*
ゲームの過程を見るときに使うファイル。
1手ごとにボードが表示されて、どこに打ったかが表示される。
rogicの部分をplayerに変更すれば対戦もできる。

コンパイル方法
> gcc simulate.c book_evaluation.c logic.c reversi.c utility.c -lm
*/

#include <stdio.h>
#include "reversi.h"
#include "logic.h"
#include <time.h>
#include <stdlib.h>
#include <sys/time.h>

int main(void) {
    Game game;
    game_init(&game);
    int pos[2];
    
    // ゲームで双方のプレイヤーが使うロジックを突っ込む。
    void (*rogic[PLAYERS])(struct Game*, int []);
    rogic[BLACK] = bot_alpha_beta;
    rogic[WHITE] = bot_softmax;

    srand(12346);

    struct timeval before, after;
    double elapse;
    long int sec, usec;

    while (TRUE) {
        // show_board(&game);

        // 時間計測
        gettimeofday(&before, NULL);
        rogic[game.turn](&game, pos);
        gettimeofday(&after, NULL);

        if (before.tv_usec > after.tv_usec) {
            // 繰り下がり
            sec = after.tv_sec - before.tv_sec - 1;
            usec = 1000000 + after.tv_usec - before.tv_usec;
        } else {
            // 何もなし
            sec = after.tv_sec - before.tv_sec;
            usec = after.tv_usec - before.tv_usec;
        }

        if (game.turn == WHITE) {
            // printf("%ld.%ld -> %ld.%ld\n", before.tv_sec, before.tv_usec, after.tv_sec, after.tv_usec);
            printf("%ld.%ld\n", sec, usec);
        }
        // printf("press -> (%d(%d), %c(%d))\n", pos[0], pos[0], pos[1] + 'A' - 1, pos[1]);
        put_stone(&game, pos, game.turn);

        if (next_turn(&game) == TRUE) {
            break;
        }
    }

    printf("game set!\n");
    show_board(&game);

    
    return 0;
}